dask/array/creation.py

```diff
@@ -721,72 +721,102 @@ def expand_pad_value(array, pad_value):
-                ith_pad_chunk_width[1] = pad_width[d][1]
-
-            ith_pad_chunk_width = tuple(ith_pad_chunk_width)
-            pad_chunk_width.append(ith_pad_chunk_width)
-
-        pad_chunk_width = tuple(pad_chunk_width)
-
-        array_chunk_key = (array.name,) + idx
-        result_chunk_key = (name,) + idx
-
-        if any(map(any, pad_chunk_width)):
-            dsk[result_chunk_key] = (
-                np_pad, array_chunk_key, pad_chunk_width, mode
-            )
-            dsk[result_chunk_key] += args
-        else:
-            dsk[result_chunk_key] = array_chunk_key
+    dtype = np.dtype(start.dtype)
+
+    result = np.empty(shape, dtype=dtype)
+    for i in np.ndindex(start.shape):
+        j = list(i)
+        j[dim] = slice(None)
+        j = tuple(j)
+
+        result[j] = np.linspace(start[i], stop, num1, dtype=dtype)[1:][::step]
+
+    return result
 
-    dsk = sharedict.merge((name, dsk), array.dask, dependencies={name: {array.name}})
 
-    result = Array(dsk, name, chunks=chunks, dtype=array.dtype)
+def pad_edge(array, pad_width, mode, *args):
+    """
+    Helper function for padding edges.
+
+    Handles the cases where the only the values on the edge are needed.
+    """
+
+    args = tuple(expand_pad_value(array, e) for e in args)
+
+    result = array
+    for d in range(array.ndim):
+        pad_shapes, pad_chunks = get_pad_shapes_chunks(result, pad_width, (d,))
+        pad_arrays = [result, result]
+
+        if mode is "constant":
+            constant_values = args[0][d]
+            constant_values = [
+                asarray(c).astype(result.dtype) for c in constant_values
+            ]
+
+            pad_arrays = [
+                broadcast_to(v, s, c)
+                for v, s, c in zip(constant_values, pad_shapes, pad_chunks)
+            ]
+        elif mode in ["edge", "linear_ramp"]:
+            pad_slices = [
+                result.ndim * [slice(None)], result.ndim * [slice(None)]
+            ]
+            pad_slices[0][d] = slice(None, 1, None)
+            pad_slices[1][d] = slice(-1, None, None)
+            pad_slices = [tuple(sl) for sl in pad_slices]
+
+            pad_arrays = [result[sl] for sl in pad_slices]
+
+            if mode is "edge":
+                pad_arrays = [
+                    broadcast_to(a, s, c)
+                    for a, s, c in zip(pad_arrays, pad_shapes, pad_chunks)
+                ]
+            elif mode is "linear_ramp":
+                end_values = args[0][d]
+
+                pad_arrays = [
+                    a.map_blocks(
+                        linear_ramp_chunk, ev, pw,
+                        chunks=c, dtype=result.dtype, dim=d, step=(2 * i - 1)
+                    )
+                    for i, (a, ev, pw, c) in enumerate(
+                        zip(pad_arrays, end_values, pad_width[d], pad_chunks)
+                    )
+                ]
+
+        result = concatenate([pad_arrays[0], result, pad_arrays[1]], axis=d)
```

Comment:
I guess your algorithm is calling concatenate on each axis right? I guess you are just setting up the computation graph. The cool things you can do with compilation!

Reply:
don't be too hung up on numpy's implementation. It was optimized for a different usecase.

Reply:
This was basically the easiest way to write it in my mind at least.

Reply:
sure, that i believe.